### PR TITLE
Updates on the parser

### PR DESCRIPTION
**Pull request by jaytaph**
This updates the parser so more tests are functioning. It still needs more work though

```
🏃‍♂️ Running 112 tests from 🗄️ "tests/data/html5lib-tests/tree-construction/tests1.dat"

🧪 Running test #1: tests/data/html5lib-tests/tree-construction/tests1.dat:1
✅ | <html>
✅ |   <head>
✅ |   <body>
✅ |     "Test"
✅ Found 1 errors
----------------------------------------
🧪 Running test #2: tests/data/html5lib-tests/tree-construction/tests1.dat:10
✅ | <html>
✅ |   <head>
✅ |   <body>
✅ |     <p>
✅ |       "One"
✅ |     <p>
✅ |       "Two"
✅ Found 1 errors
----------------------------------------
🧪 Running test #3: tests/data/html5lib-tests/tree-construction/tests1.dat:23
✅ | <html>
✅ |   <head>
✅ |   <body>
✅ |     "Line1"
✅ |     <br>
✅ |     "Line2"
✅ |     <br>
✅ |     "Line3"
✅ |     <br>
✅ |     "Line4"
✅ Found 1 errors
----------------------------------------
🧪 Running test #4: tests/data/html5lib-tests/tree-construction/tests1.dat:39
✅ | <html>
✅ Found 1 errors
----------------------------------------
🧪 Running test #5: tests/data/html5lib-tests/tree-construction/tests1.dat:48
✅ | <html>
✅ |   <head>
✅ Found 1 errors
----------------------------------------
🧪 Running test #6: tests/data/html5lib-tests/tree-construction/tests1.dat:57
✅ | <html>
✅ |   <head>
✅ |   <body>
✅ Found 1 errors
----------------------------------------
🧪 Running test #7: tests/data/html5lib-tests/tree-construction/tests1.dat:66
✅ | <html>
✅ |   <head>
✅ Found 1 errors
----------------------------------------
🧪 Running test #8: tests/data/html5lib-tests/tree-construction/tests1.dat:75
✅ | <html>
✅ |   <head>
✅ Found 1 errors
----------------------------------------
🧪 Running test #9: tests/data/html5lib-tests/tree-construction/tests1.dat:84
✅ | <html>
✅ |   <head>
✅ |   <body>
✅ Found 1 errors
----------------------------------------
🧪 Running test #10: tests/data/html5lib-tests/tree-construction/tests1.dat:93
✅ | <html>
✅ |   <head>
✅ |   <body>
✅ Found 1 errors
----------------------------------------
🧪 Running test #11: tests/data/html5lib-tests/tree-construction/tests1.dat:102
✅ | <html>
✅ |   <head>
✅ |   <body>
✅ Found 1 errors
----------------------------------------
🧪 Running test #12: tests/data/html5lib-tests/tree-construction/tests1.dat:111
✅ | <html>
✅ |   <head>
✅ |   <body>
✅ Found 1 errors
----------------------------------------
🧪 Running test #13: tests/data/html5lib-tests/tree-construction/tests1.dat:120
✅ | <html>
✅ |   <head>
✅ |   <body>
✅ Found 1 errors
----------------------------------------
🧪 Running test #14: tests/data/html5lib-tests/tree-construction/tests1.dat:129
✅ | <html>
✅ |   <head>
✅ |   <body>
✅ Found 1 errors
----------------------------------------
🧪 Running test #15: tests/data/html5lib-tests/tree-construction/tests1.dat:138
✅ | <html>
✅ |   <head>
✅ |   <body>
✅ Found 1 errors
----------------------------------------
🧪 Running test #16: tests/data/html5lib-tests/tree-construction/tests1.dat:147
✅ | <html>
✅ |   <head>
✅ |   <body>
✅ Found 1 errors
----------------------------------------
🧪 Running test #17: tests/data/html5lib-tests/tree-construction/tests1.dat:156
✅ | <html>
✅ |   <head>
✅ Found 1 errors
----------------------------------------
🧪 Running test #18: tests/data/html5lib-tests/tree-construction/tests1.dat:165
✅ | <html>
✅ |   <head>
✅ |   <body>
✅ Found 1 errors
----------------------------------------
🧪 Running test #19: tests/data/html5lib-tests/tree-construction/tests1.dat:174
✅ | <html>
✅ |   <head>
✅ |   <body>
✅ Found 1 errors
----------------------------------------
🧪 Running test #20: tests/data/html5lib-tests/tree-construction/tests1.dat:183
✅ | <html>
✅ |   <head>
✅ |   <body>
✅ |     <b>
✅ |       <table>
✅ |         <tbody>
✅ |           <tr>
✅ |             <td>
✅ |               <i>
❌ Unexpected errors found (wanted 4, got 3): 
     * Want: 'expected-doctype-but-got-start-tag' at 1:3
     * Want: 'unexpected-cell-in-table-body' at 1:14
     * Want: 'unexpected-cell-end-tag' at 1:25
     * Want: 'expected-closing-tag-but-got-eof' at 1:25
     * Got: 'expected-doctype-but-got-start-tag' at 1:4
     * Got: 'th or td tag not allowed in in table body insertion mode' at 1:15
     * Got: 'current node should be td or th' at 1:26
----------------------------------------
📄 Input stream: 
<b><table><td><i></table>
----------------------------------------
🌳 Generated tree: 
└─ Document
   └─ <html>
      ├─ <head>
      └─ <body>
         └─ <b>
            └─ <table>
               └─ <tbody>
                  └─ <tr>
                     └─ <td>
                        └─ <i>

----------------------------------------
🌳 Expected tree: 
| <html>
|   <head>
|   <body>
|     <b>
|       <table>
|         <tbody>
|           <tr>
|             <td>
|               <i>

----------------------------------------
🧪 Running test #21: tests/data/html5lib-tests/tree-construction/tests1.dat:201
✅ | <html>
✅ |   <head>
✅ |   <body>
✅ |     <b>
✅ |       <table>
✅ |         <tbody>
✅ |           <tr>
✅ |             <td>
✅ |               <i>
❌ Unexpected errors found (wanted 5, got 4): 
     * Want: 'expected-doctype-but-got-start-tag' at 1:3
     * Want: 'unexpected-cell-in-table-body' at 1:14
     * Want: 'unexpected-end-tag' at 1:18
     * Want: 'unexpected-cell-end-tag' at 1:29
     * Want: 'expected-closing-tag-but-got-eof' at 1:30
     * Got: 'expected-doctype-but-got-start-tag' at 1:4
     * Got: 'th or td tag not allowed in in table body insertion mode' at 1:15
     * Got: 'special node' at 1:19
     * Got: 'current node should be td or th' at 1:30
----------------------------------------
📄 Input stream: 
<b><table><td></b><i></table>X
----------------------------------------
🌳 Generated tree: 
└─ Document
   └─ <html>
      ├─ <head>
      └─ <body>
         └─ <b>
            └─ <table>
               └─ <tbody>
                  └─ <tr>
                     └─ <td>
                        └─ <i>

----------------------------------------
🌳 Expected tree: 
| <html>
|   <head>
|   <body>
|     <b>
|       <table>
|         <tbody>
|           <tr>
|             <td>
|               <i>
|       "X"

----------------------------------------
🧪 Running test #22: tests/data/html5lib-tests/tree-construction/tests1.dat:221
✅ | <html>
✅ |   <head>
✅ |   <body>
✅ |     <h1>
✅ |       "Hello"
✅ |     <h2>
✅ |       "World"
❌ Unexpected errors found (wanted 3, got 2): 
     * Want: 'expected-doctype-but-got-start-tag' at 1:4
     * Want: 'unexpected-start-tag' at 1:13
     * Want: 'expected-closing-tag-but-got-eof' at 1:18
     * Got: 'expected-doctype-but-got-start-tag' at 1:5
     * Got: 'h1-h6 not allowed in in body insertion mode' at 1:14
----------------------------------------
📄 Input stream: 
<h1>Hello<h2>World
----------------------------------------
🌳 Generated tree: 
└─ Document
   └─ <html>
      ├─ <head>
      └─ <body>
         ├─ <h1>
         │  └─ "Hello"
         └─ <h2>
            └─ "World"

----------------------------------------
🌳 Expected tree: 
| <html>
|   <head>
|   <body>
|     <h1>
|       "Hello"
|     <h2>
|       "World"

----------------------------------------
🧪 Running test #23: tests/data/html5lib-tests/tree-construction/tests1.dat:236

```

### CODE REVIEW

**Comment by emwalker** [comment]
Looks like there's a failing test.  I'd either try to fix the test or add `#[ignore]` to the test if this should be merged without fixing it.  Allowing test regressions at this stage doesn't seem like a big deal, so I wouldn't personally worry about disabling a test.

**Comment by emwalker** [comment]
@jaytaph how about these further changes?  After I do this, the tests pass.

```patch
diff --git a/src/html5_parser/parser/adoption_agency.rs b/src/html5_parser/parser/adoption_agency.rs
index 6b7ffb7..c2d9e00 100755
--- a/src/html5_parser/parser/adoption_agency.rs
+++ b/src/html5_parser/parser/adoption_agency.rs
@@ -323,6 +323,7 @@ mod test {
         }};
     }
 
+    #[ignore]
     #[test]
     fn test_find_furthest_block_idx_1() {
         let mut stream = InputStream::new();
@@ -383,6 +384,7 @@ mod test {
         );
     }
 
+    #[ignore]
     #[test]
     fn test_find_furthest_block_idx_2() {
         let mut stream = InputStream::new();
```

File: src/html5_parser/node.rs [hunk 1]
```diff
@@ -33,8 +34,8 @@ pub enum NodeData {
 }
 
 /// Id used to identify a node
-#[derive(Clone, Copy, Debug, Default, Eq, Hash, PartialEq)]
-pub struct NodeId(usize);
+#[derive(Clone, Copy, Debug, Default, Eq, Hash, PartialEq, Display)]
+pub struct NodeId(pub usize);
```
Comment (emwalker) [inline src/html5_parser/node.rs:38]: Making the internal `usize` is probably not needed, but I think it's fine for now.  This kind of thing only matters when external libraries are depending on this code.

Comment (jaytaph) [reply src/html5_parser/node.rs:38]: I needed it for using them inside unit-tests. Maybe we can make this better by using a `NodeId::from("5")` method?

Comment (blaumeise20) [reply src/html5_parser/node.rs:38]: You can say `pub(crate) usize` to make it only visible to the current crate.

**Pull request merged**